File: server/src/collectors/osmTransportSubways.js
/**
 * OSM Transport Layer — Subway / Metro / Monorail / Tram (always-on)
 *
 * Dedicated nationwide Overpass pull of `station=subway`, `station=light_rail`,
 * `station=monorail`, and `railway=tram_stop`. Complements ODPT (which covers
 * Tokyo Metro + Toei + a handful of paid operators) with full OSM coverage
 * for Osaka, Nagoya, Sapporo, Sendai, Fukuoka, Kyoto, Kobe, Yokohama metros
 * as well as the Okinawa monorail, tram networks, and people-movers.
 */

import { fetchOverpassTiled } from './_liveHelpers.js';

export default async function collectOsmTransportSubways() {
  const features = await fetchOverpassTiled(
    (bbox) => [
      `node["station"="subway"](${bbox});`,
      `node["station"="light_rail"](${bbox});`,
      `node["station"="monorail"](${bbox});`,
      `node["railway"="tram_stop"](${bbox});`,
      `node["public_transport"="station"]["subway"="yes"](${bbox});`,
      `node["public_transport"="station"]["monorail"="yes"](${bbox});`,
      `node["public_transport"="station"]["tram"="yes"](${bbox});`,
    ].join(''),
    (el, _i, coords) => ({
      type: 'Feature',
      geometry: { type: 'Point', coordinates: coords },
      properties: {
        station_id: `OSM_${el.id}`,
        name: el.tags?.['name:en'] || el.tags?.name || 'Station',
        name_ja: el.tags?.name || el.tags?.['name:ja'] || null,
        line: el.tags?.line || el.tags?.network || null,
        operator: el.tags?.operator || null,
        type: el.tags?.station || el.tags?.railway || 'subway',
        network: el.tags?.network || null,
        wikidata: el.tags?.wikidata || null,
        wheelchair: el.tags?.wheelchair || null,
        country: 'JP',
        source: 'osm_transport_subways',
      },
    }),
    { queryTimeout: 180, timeoutMs: 120_000 },
  );

  const list = features || [];
  return {
    type: 'FeatureCollection',
    features: list,
    _meta: {
      source: 'osm_transport_subways',
      fetchedAt: new Date().toISOString(),
      recordCount: list.length,
      live: list.length > 0,
      description: 'OSM always-on layer for subway / metro / monorail / tram / light_rail stops across Japan',
    },
    metadata: {},
  };
}

File: native/collectors/sources/osm_transport_subways.c
/* collectors/transport/sources/osm_transport_subways.c — port of
 * server/src/collectors/osmTransportSubways.js
 * (createOsmTransportCollector, geometry:'point' → fetchOverpassTiled).
 * Always-on subway / metro / monorail / tram / light_rail stops.
 * No SEED / _meta envelope. */
#include "../../lib/geojson.h"
#include "../../source.h"
#include "../../lib/overpass.h"
#include "../../lib/linecolor.h"
#include <stdio.h>

static void body(const char *bbox, char *o, size_t n, void *ud) {
  (void)ud;
  snprintf(o, n,
    "node[\"station\"=\"subway\"](%s);"
    "node[\"station\"=\"light_rail\"](%s);"
    "node[\"station\"=\"monorail\"](%s);"
    "node[\"railway\"=\"tram_stop\"](%s);"
    "node[\"public_transport\"=\"station\"][\"subway\"=\"yes\"](%s);"
    "node[\"public_transport\"=\"station\"][\"monorail\"=\"yes\"](%s);"
    "node[\"public_transport\"=\"station\"][\"tram\"=\"yes\"](%s);",
    bbox, bbox, bbox, bbox, bbox, bbox, bbox);
}

/* properties built in EXACT JS key order. */
static cJSON *map(cJSON *el, int i, double lon, double lat, void *ud) {
  (void)i; (void)ud;
  cJSON *f = gj_point_feature(lon, lat);

  cJSON *p = cJSON_CreateObject();
  cJSON *idv = cJSON_GetObjectItem(el, "id");
  char sid[64];
  snprintf(sid, sizeof sid, "OSM_%lld",
           idv && cJSON_IsNumber(idv) ? (long long)idv->valuedouble : 0);
  cJSON_AddStringToObject(p, "station_id", sid);

  const char *nm = ov_tag(el, "name:en"); if (!nm) nm = ov_tag(el, "name");
  cJSON_AddStringToObject(p, "name", nm ? nm : "Station");
  const char *nj = ov_tag(el, "name"); if (!nj) nj = ov_tag(el, "name:ja");
  cJSON_AddItemToObject(p, "name_ja", nj ? cJSON_CreateString(nj) : cJSON_CreateNull());
  const char *ln = ov_tag(el, "line"); if (!ln) ln = ov_tag(el, "network");
  cJSON_AddItemToObject(p, "line", ln ? cJSON_CreateString(ln) : cJSON_CreateNull());
  const char *op = ov_tag(el, "operator");
  cJSON_AddItemToObject(p, "operator", op ? cJSON_CreateString(op) : cJSON_CreateNull());
  const char *st = ov_tag(el, "station"); if (!st) st = ov_tag(el, "railway");
  cJSON_AddStringToObject(p, "type", st ? st : "subway");
  const char *nw = ov_tag(el, "network");
  cJSON_AddItemToObject(p, "network", nw ? cJSON_CreateString(nw) : cJSON_CreateNull());
  const char *wd = ov_tag(el, "wikidata");
  cJSON_AddItemToObject(p, "wikidata", wd ? cJSON_CreateString(wd) : cJSON_CreateNull());
  const char *wc = ov_tag(el, "wheelchair");
  cJSON_AddItemToObject(p, "wheelchair", wc ? cJSON_CreateString(wc) : cJSON_CreateNull());
  char lc[8];
  cJSON_AddItemToObject(p, "line_color",
    line_color(cJSON_GetObjectItem(el, "tags"), lc) ? cJSON_CreateString(lc)
                                                    : cJSON_CreateNull());
  cJSON_AddStringToObject(p, "country", "JP");
  cJSON_AddStringToObject(p, "source", "osm_transport_subways");
  cJSON_AddItemToObject(f, "properties", p);
  return f;
}

static int run(const source_ctx *ctx, intel_sink *sink) {
  int n = overpass_tiled_collect(ctx, sink, body, 180, 120000, map, NULL);
  return n >= 0 ? 0 : -1;
}

static const source_def osm_transport_subways_def = {
  .id = "osm-transport-subways", .collector = "transport",
  .name = "OSM Transport Subways", .name_ja = "OSM Subways",
   .update_interval_sec = 86400, .run = run };
REGISTER_SOURCE(osm_transport_subways_def)
